### ISSUE TITLE
Allow more customization: e.g. for input field

### ISSUE THREAD

**Issue by jangruenwaldt** [flyerhq/flutter_chat_ui#22]
A really simple change: if the user provides a 'WidgetBuilder inputBuilder' argument, that is used to render the input. Makes it easier to reuse and restyle for your project.

Thoughts?

**Comment by demchenkoalex** [comment]
Hi @jangruenwaldt yes, simple to add, but we need to think about how to connect it to the message user is typing inside the `inputBuilder`, to the send and attachment buttons etc. You can now customise colors, border radius (AFAIR, not sure), icons.. How the input may be so different it needs its own builder? But for messages yeah, we should add those.

**Comment by jangruenwaldt** [comment]
Good point yeah! 

For me it was like this: we have our own CustomTextField, that we always use to render inputs - rebuilding that in a custom theme is quite cumbersome so I added a builder function in my fork. Atleast for me thats what I always prefer, since my app already has widgets and a theme and everything, so i dont really want to rebuild that for each package I use, but instead just provide a custom builder function that builds it according to my needs. Also, for example if I want to use e.g. a speed dial for the attachment etc I can not do that currently (https://pub.dev/packages/flutter_speed_dial).

But I also dont know whats best - that's just my thoughts on it - the package is so awesome already.

**Comment by demchenkoalex** [comment]
We'll think how can we improve here :) will add this task to our public board.

**Comment by jangruenwaldt** [comment]
Btw, same goes for basically anything - the read indicator etc, the message bubble itself. Its a usual pattern in flutter packages to provide a builder function to customize it.

I needed to add a message wrapper builder in my fork, so I can wrap the bubble with a popupMenuButton to properly handle onTap events.

Much easier also since there is no need for two seperate arguments - onTap and onLongPress. Simply one wrapperBuilder argument.

Would be good for everyone if more of those are added, but as always, no worries and no pressure, I know there are more important things right now.

**Comment by fxjordan** [comment]
Hi, +1 for a custom builder for individual messages.

I need to implement a custom message type that references a user post in our app. Any ideas how to accomplish this at the moment?

**Comment by demchenkoalex** [comment]
Hi @fxjordan we have 3 types of messages right now, text, image and file. If it's not possible to do with one of them, then not possible at all. After we release 1.1 (pinned issue) I think I will be able to do a custom message and custom message widget prop to render whatever is needed.

**Comment by avinashk369** [comment]
i would like to add couple of ui elements to make input field more customisable 


![Simulator Screen Shot - iPhone SE (2nd generation) - 2021-08-31 at 17 49 54](https://user-images.githubusercontent.com/5073425/131502042-cc6c73b2-8f7f-4579-9e16-6729152bd1f4.png)

# **default messageBox UI**


![Simulator Screen Shot - iPhone SE (2nd generation) - 2021-08-31 at 17 51 26](https://user-images.githubusercontent.com/5073425/131502051-aaa0efde-eb4e-46e9-99d1-c23c3bce9d90.png)

# **custom messageBox UI**

added three new params in ChatTheme

```
this.messageBoxMargin = 0.0,
this.cursorColor = NEUTRAL_0,
this.inputBorder = InputBorder.none,
```

and used them inside focus widget of Input class

```
 Focus(
            autofocus: true,
            child: Padding(
              padding: EdgeInsets.all(
                  InheritedChatTheme.of(context).theme.messageBoxMargin),
              child: Material(
                borderRadius:
                    InheritedChatTheme.of(context).theme.inputBorderRadius,
                color:
                    InheritedChatTheme.of(context).theme.inputBackgroundColor,
                child: Container(
                  padding: EdgeInsets.fromLTRB(
                    24 + _query.padding.left,
                    20,
                    24 + _query.padding.right,
                    20 + _query.viewInsets.bottom + _query.padding.bottom,
                  ),
                  child: Row(
                    children: [
                      if (widget.onAttachmentPressed != null) _leftWidget(),
                      Expanded(
                        child: TextField(
                          controller: _textController,
                          decoration: InputDecoration(
                            isCollapsed: true,
                            border: InheritedChatTheme.of(context)
                                .theme
                                .inputBorder,
                            hintStyle: InheritedChatTheme.of(context)
                                .theme
                                .inputTextStyle
                                .copyWith(
                                  color: InheritedChatTheme.of(context)
                                      .theme
                                      .inputTextColor
                                      .withOpacity(0.5),
                                ),
                            hintText:
                                InheritedL10n.of(context).l10n.inputPlaceholder,
                          ),
                          focusNode: _inputFocusNode,
                          keyboardType: TextInputType.multiline,
                          maxLines: 5,
                          cursorColor:
                              InheritedChatTheme.of(context).theme.cursorColor,
                          minLines: 1,
                          onChanged: widget.onTextChanged,
                          style: InheritedChatTheme.of(context)
                              .theme
                              .inputTextStyle
                              .copyWith(
                                color: InheritedChatTheme.of(context)
                                    .theme
                                    .inputTextColor,
                              ),
                          textCapitalization: TextCapitalization.sentences,
                        ),
                      ),
                      Visibility(
                        visible: _sendButtonVisible,
                        child: SendButton(
                          onPressed: _handleSendPressed,
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
          )
```

inside focus widget had to replace InputDecoration.collapsed with InputDecoration.

this is small changes but makes real difference.

**Comment by demchenkoalex** [comment]
Released in [v1.3.2](https://github.com/flyerhq/flutter_chat_ui/releases/tag/v1.3.2). As for the chat bubbles customisation, this ticket has become a mess with all kind of different requests, so I am creating a [discussion](https://github.com/flyerhq/flutter_chat_ui/discussions/102) for the chat bubbles, with on option to upvote, but anyway this will probably be the next feature we will do.

**Comment by jader13254** [comment]
Appreciate the changes, just wondering, have you ever considered using a builder pattern? That's quite common, I feel like it would be helpful for flyer chat. Since these nrw options can never cover everything.

E.g. we forked flyer chat and heavily customized everything using builders. Added a button for polls etc. And pollMessageType...

It would be awesome if you could - without forking -e.g.  add a new message type + builder function. It would be doable, with a improved structure using builders.

**Comment by demchenkoalex** [comment]
Hi @jader13254 I learned Flutter while doing this, I am an iOS dev and working in React Native. Somebody already mentioned a builder pattern and even sent some examples, and I didn't understand it at all (maybe examples were to complicated to start with). Better to have something forkable than nothing I guess.. Instead of builders I have added a custom message type which has a map for anything you want it to have and a custom message builder (not a builder pattern I guess) which can build anything you want. But really the point was to create a library with all message types you would see in a normal chat app, including audio, video, stickers etc, and make it work with as few lines of code as possible. I have never planned to build a monstrous "customize everything" or "build complete app" without putting any effort. If somebody requires many different message types it will be 100x faster to build it from scratch than using any available package on any platform IMO. If it's like one or two, my understanding that custom message builder is enough.
